native/dlaqr5.go

```diff
@@ -0,0 +1,609 @@
+			var jtop int
+			switch {
+			case accum:
+				jtop = max(ktop, incol)
+			case wantt:
+				jtop = 0
+			default:
+				jtop = ktop
+			}
+			for m := mtop; m <= mbot; m++ {
+				if v[m*ldv] == 0 {
+					continue
+				}
+				k := krcol + 3*m
+				for j := jtop; j <= min(kbot, k+3); j++ {
+					refsum := v[m*ldv] * (h[j*ldh+k+1] + v[m*ldv+1]*h[j*ldh+k+2] + v[m*ldv+2]*h[j*ldh+k+3])
+					h[j*ldh+k+1] -= refsum
+					h[j*ldh+k+2] -= refsum * v[m*ldv+1]
+					h[j*ldh+k+3] -= refsum * v[m*ldv+2]
+				}
+				if accum {
+					// Accumulate U. (If necessary, update Z later with with an
+					// efficient matrix-matrix multiply.)
+					kms := k - incol
+					for j := max(1, ktop-incol) - 1; j < kdu; j++ {
+						refsum := v[m*ldv] * (u[j*ldu+kms] + v[m*ldv+1]*u[j*ldu+kms+1] + v[m*ldv+2]*u[j*ldu+kms+2])
+						u[j*ldu+kms] -= refsum
+						u[j*ldu+kms+1] -= refsum * v[m*ldv+1]
+						u[j*ldu+kms+2] -= refsum * v[m*ldv+2]
+					}
+				} else if wantz {
+					// U is not accumulated, so update Z now by multiplying by
+					// reflections from the right.
+					for j := iloz; j <= ihiz; j++ {
+						refsum := v[m*ldv] * (z[j*ldz+k+1] + v[m*ldv+1]*z[j*ldz+k+2] + v[m*ldv+2]*z[j*ldz+k+3])
+						z[j*ldz+k+1] -= refsum
+						z[j*ldz+k+2] -= refsum * v[m*ldv+1]
+						z[j*ldz+k+3] -= refsum * v[m*ldv+2]
+					}
+				}
+			}
+
+			// Special case: 2×2 reflection (if needed).
+			if bmp22 && v[m22*ldv] != 0 {
+				k := krcol + 3*m22
+				for j := jtop; j <= min(kbot, k+3); j++ {
+					refsum := v[m22*ldv] * (h[j*ldh+k+1] + v[m22*ldv+1]*h[j*ldh+k+2])
+					h[j*ldh+k+1] -= refsum
+					h[j*ldh+k+2] -= refsum * v[m22*ldv+1]
+				}
+				if accum {
+					kms := k - incol
+					for j := max(1, ktop-incol) - 1; j < kdu; j++ {
+						refsum := v[m22*ldv] * (u[j*ldu+kms] + v[m22*ldv+1]*u[j*ldu+kms+1])
+						u[j*ldu+kms] -= refsum
+						u[j*ldu+kms+1] -= refsum * v[m22*ldv+1]
+					}
+				} else if wantz {
+					for j := iloz; j <= ihiz; j++ {
+						refsum := v[m22*ldv] * (z[j*ldz+k+1] + v[m22*ldv+1]*z[j*ldz+k+2])
+						z[j*ldz+k+1] -= refsum
+						z[j*ldz+k+2] -= refsum * v[m22*ldv+1]
+					}
+				}
+			}
+
+			// Vigilant deflation check.
+			mstart := mtop
+			if krcol+3*mstart < ktop {
+				mstart++
+			}
+			mend := mbot
+			if bmp22 {
+				mend++
+			}
+			if krcol == kbot-2 {
+				mend++
+			}
+			for m := mstart; m <= mend; m++ {
+				k := min(kbot-1, krcol+3*m)
+
+				// The following convergence test requires that the tradition
+				// small-compared-to-nearby-diagonals criterion and the Ahues &
+				// Tisseur (LAWN 122, 1997) criteria both be satisfied. The latter
+				// improves accuracy in some examples. Falling back on an alternate
+				// convergence criterion when tst1 or tst2 is zero (as done here) is
+				// traditional but probably unnecessary.
+
+				if h[(k+1)*ldh+k] == 0 {
+					continue
+				}
+				tst1 := math.Abs(h[k*ldh+k]) + math.Abs(h[(k+1)*ldh+k+1])
+				if tst1 == 0 {
+					if k >= ktop+1 {
```

Comment:
I don't see how to convert this to a switch easily. All the if conditions can be true.